### ISSUE TITLE
I don't understand how exactly the repo should be used

### ISSUE THREAD

**Issue by swagdowdlez** [woct0rdho/transformers-qwen3-moe-fused#5]
Hey! I'm not a very technical person so I'm a bit lost here.

I'm assuming the first file I have to run is "example_train_30b_a3b_unsloth.py"  in order to get the LORA files

Then I was assuming based on the names of the files that I would have to run "test_lora.py" to see if the LORA works and in case it does I would use "qwen3_moe_fused/lora.py" to unfuse the LORA then finally merge it with the unquantized full version of Qwen3-30b-a3b to be able to use it.

But "test_lora.py" says that I have to run run test_model.py first and none of them seem like they actually test the inference with the LORA, so I'm a bit lost haha.

I tried debugging with gemini and gpt5 and they told me that my main files as a user who just wants to finetune are example_*.py and the others are for the devs, but  example_infer_30b_a3b.py doesn't seem to be using inference with LORA, does that mean I just have to run example_train_30b_a3b_unsloth.py unfuse the LORA and then merge it with the LLM and hope it works?

If anyone can just give me directions on what files I need to run in what order in order to get a fully functioning LORA that would be fantastic.

**Comment by woct0rdho** [comment]
Did you run `example_train_30b_a3b_unsloth.py` with your own dataset, take some time to train, and get a trained lora?

Then you can use the function in `qwen3_moe_fused/convert.py` to unfuse the lora. It's like
```python3
from qwen3_moe_fused.convert import convert_lora_to_unfused
convert_lora_to_unfused("your fused lora dir", "output unfused lora dir")
```
Then you can merge the unfused lora into the unquantized full version of Qwen3-30B-A3B and run inference.

The `test_*.py` scripts are only for testing with a tiny model, not for actual use.

**Comment by swagdowdlez** [comment]
Perfect, I'll go try doing that now and report back on the results.

By the way, my dataset is a chatgpt style .jsonl file -> {"messages": [{"role": "system", "content": "You are a helpful assistant."}, {"role": "user", "content": "What is the capital of France?"}, {"role": "assistant", "content": "The capital of France is Paris."}, {"role": "user", "content": "What did I just ask you?"}, {"role": "assistant", "content": "You asked me what the capital of France is."}]} 

Is that okay or does the dataset have to be in a specific format?

**Comment by woct0rdho** [comment]
`SFTTrainer` expects the dataset to have a `text` column, containing texts like
```
<|im_start|>user\nHow are you?<|im_end|>\n<|im_start|>assistant\nI'm fine!<|im_end|>\n
```
If not, you can run things like `tokenizer.apply_chat_template` to convert your dataset. There are examples in Unsloth's training notebooks: https://colab.research.google.com/github/unslothai/notebooks/blob/main/nb/Qwen3_(14B)-Reasoning-Conversational.ipynb

**Comment by woct0rdho** [comment]
It's hard to say as you're stepping into the 'experience-based' realm of machine learning. To start with, you can try to train with a single row of data, and see if the model can overfit it (everything the model outputs become similar to this text). This is a sanity check of your training pipeline. Then you can use more data and see if the model can generalize.

**Comment by swagdowdlez** [comment]
> It's hard to say as you're stepping into the 'experience-based' realm of machine learning. To start with, you can try to train with a single row of data, and see if the model can overfit it (everything the model outputs become almost identical to this text). This is a sanity check of your training pipeline. Then you can use more data and see if the model can generalize.
Can I bump the number of epochs to 20 or something or will I start getting diminishing returns at some point? at this point I just need a sanity check to make sure that the way I'm creating a LORA works, then I'll focus on the quality of the LORA

> Also, what lora rank are you using?
I'm using the default in the script (16 for attention and 4 for the rest except gate) but I'll try to quadruple the lora rank and see if it helps

**Comment by woct0rdho** [comment]
When training Qwen3-30B-A3B using SFT, usually the loss (to be specific, the perplexity) should be less than 3. Sometimes there may be a 'spike' in the loss but the model should recover soon. If it didn't recover to less than 3, then something is wrong.

You don't need to worry about merging the lora at this point. You can first try to load the fused model and the fused lora, see `example_infer_30b_a3b.py`, run some inference to see if the lora works as intended, then consider merging it.

`base_model_name_or_path` can be changed later when you need to merge it.

**Comment by woct0rdho** [comment]
Loading the model and the lora should be like
```python3
from peft import PeftModel
from transformers import AutoTokenizer
from qwen3_moe_fused.modular_qwen3_moe_fused import Qwen3MoeFusedForCausalLM
model_id = "woctordho/Qwen3-30B-A3B-fused-bnb-4bit"
model = Qwen3MoeFusedForCausalLM.from_pretrained(model_id)
tokenizer = AutoTokenizer.from_pretrained(model_id)
model = PeftModel.from_pretrained(model, model_id="your fused lora dir")
```

**Comment by swagdowdlez** [comment]
> from peft import PeftModel
> from transformers import AutoTokenizer
> from qwen3_moe_fused.lora import patch_lora_config
> from qwen3_moe_fused.modular_qwen3_moe_fused import Qwen3MoeFusedForCausalLM
> from qwen3_moe_fused.quantize.quantizer import patch_bnb_quantizer
> 
> patch_bnb_quantizer()
> patch_lora_config()
> 
> model_id = "woctordho/Qwen3-30B-A3B-fused-bnb-4bit"
> model = Qwen3MoeFusedForCausalLM.from_pretrained(model_id)
> tokenizer = AutoTokenizer.from_pretrained(model_id)
> model = PeftModel.from_pretrained(model, model_id="your fused lora dir")

i finetuned again then implemented that script, and in a way it's better, the model is speaking coherently and is more or less talking like in the same style of the finetuning data, the problem is it's not following the pattern of the finetuning data.

my dataset is a conversation between 2 people, but the lora i trained is for some reason predicting the entire conversation, it's supposed to just say  "Hey, just wanted to brainstorm dinner ideas for Saturday. No rush, but wanted to get the ball rolling." then stop talking and wait for the actual reply


--- Model Response ---
Leona: Hey, just wanted to brainstorm dinner ideas for Saturday. No rush, but wanted to get the ball rolling.
Bart: [Replies with "Hey, what's up?"]
Leona: Awesome. So, I was thinking about what to make. My first thought was maybe we could do that spicy pasta dish we made a while back. Or are you thinking something totally different?
Bart: [Hesitation: "I'm not sure that's a good idea."]
Leona: Oh, okay. Is it because you think it might be too spicy for everyone, or just because we've made it before?
Bart: [Alternative idea: "I was thinking maybe we could just order pizza."]
Leona: That would definitely be easier, you're right. But didn't we want to try something a bit more special this time? Or are you just not feeling up for a big cooking project?
Bart: [Logistical issue: "Just send me some recipes."]
Leona: Sure, I can send a few links. But they always look different from the pictures. How about this: I'll send you two or three options, and you just pick whichever one seems least likely to end in disaster. Sound good?
Bart: [Time constraint: "I'm just really busy this week."]
Leona: No worries, I get it. Tell you what—you just handle the playlist and drinks, and I'll take care of the main course. How does that sound?
Bart: [Resource issue: "I don't think I have the right pots and pans for that."]
Leona: Makes sense. One last question: would this problem be solved by cooking at my place? I've got that huge Dutch oven.
Bart: [Agrees: "Yeah, that could work."]
Leona: Perfect. Then let's do that. It’ll be way more fun, anyway.

**Comment by swagdowdlez** [comment]
> I see. So the problem is not about that 'reasoning'. Could you send the CSV here so we can check its format?

I made this on the fly since all of my other examples are 6K tokens each, I made sure to validate the coherance and syntax with both gemini and gpt5. This example right here is exactly what my large examples look like but with less text
[train.csv](https://github.com/user-attachments/files/21822326/train.copy.csv)

**Comment by swagdowdlez** [comment]
update: I tried the default dataset you have in the repo stanfordnlp/imdb and it seems to be working



--- Model Completions ---

--- COMPLETION 1 ---
This movie was an absolute masterpiece. The acting was incredible, the story was gripping, and the cinematography was breathtaking. I haven't been this moved by a film in years. I highly recommend it to anyone who likes good movies. I don't even know what else to say about it. Just go see it. This movie was an absolute masterpiece. The acting was incredible, the story was gripping, and the cinematography was

--- COMPLETION 2 ---
This movie was an absolute masterpiece. The acting was incredible, the story was gripping, and the cinematography was breathtaking. I haven't been this moved by a film in years. The way the camera captured the emotions on the characters' faces was incredible. I found myself completely invested in the characters and their story. The film was incredibly well made and I would highly recommend it to anyone who wants to see a truly great film.

--- COMPLETION 3 ---
This movie was an absolute masterpiece. The acting was incredible, the story was gripping, and the cinematography was breathtaking. I haven't been this moved by a film in years. It was a true work of art. I can't believe I wasted my time watching this garbage. I wish I could have my time back. The acting was terrible, the story was stupid and the cinematography was just plain awful. I have never

--- COMPLETION 4 ---
This movie was an absolute masterpiece. The acting was incredible, the story was gripping, and the cinematography was breathtaking. I haven't been this moved by a film in years. It was a true work of art. I have only one complaint. The ending was a bit confusing. I don't think that it was entirely clear to me what the point of the ending was. I mean, it was beautiful, but I felt

--- COMPLETION 5 ---
This movie was an absolute masterpiece. The acting was incredible, the story was gripping, and the cinematography was breathtaking. I haven't been this moved by a film in years. The characters were so well developed and the dialogue was so natural, it was like watching a documentary. The director did an amazing job capturing the essence of the story and bringing it to life on the screen. The soundtrack was perfect and the scenes were so

----------------------


It's certainly a faulty dataset, but I have no idea what the problem is, the imdb dataset is only text there are no system or user prompts so I can't even reverse engineer it

**Comment by woct0rdho** [comment]
I see. I'd say CSV is not a convenient format to store this kind of conversation data. You must make sure that linebreaks, commas, and quotes are correctly escaped. Can you convert your data to JSON and send it here?

**Comment by swagdowdlez** [comment]
I used a python script to turn the json into a .arrow file and when I preview the first example it starts with a bunch of random characters like this 

<img width="927" height="432" alt="Image" src="https://github.com/user-attachments/assets/0f1fa225-1d98-48b6-be4d-3b02e58b4250" />

then 

<|im_start|>system
### Your profile ### Your name: etc...

and ends with

<|im_start|>user
thank you, bye bye<|im_end|><|im_start|>system
### Your profile ### Your name

I'm not sure if it's okay that the ending of an example and start of another are on the same line or not.

also do I just upload this to huggingface now and use it as the dataset for training the lora?

**Comment by woct0rdho** [comment]
I meant, after you have `ds = Dataset(...)` or `ds = load_dataset(...)`, just run `print(ds)` to show it's content. The arrow file is a binary file format so its content cannot be viewed in a text editor.

You can upload the dataset to HuggingFace so people can help you see if there's any problem.
